src/fetchInterceptor.js

```diff
@@ -0,0 +1,249 @@
+      createAccessTokenRequest: null,
+      shouldIntercept: null,
+      shouldInvalidateAccessToken: null,
+      parseAccessToken: null,
+      authorizeRequest: null,
+      onAccessTokenChange: null,
+      onResponse: null,
+    };
+
+    this.isConfigValid = this.isConfigValid.bind(this);
+    this.fetchWithRetry = this.fetchWithRetry.bind(this);
+    this.intercept = this.intercept.bind(this);
+    this.resolveIntercept = this.resolveIntercept.bind(this);
+
+    this.createRequestUnit = this.createRequestUnit.bind(this);
+    this.shouldIntercept = this.shouldIntercept.bind(this);
+    this.authorizeRequest = this.authorizeRequest.bind(this);
+    this.shouldFetch = this.shouldFetch.bind(this);
+    this.fetchRequest = this.fetchRequest.bind(this);
+    this.shouldInvalidateAccessToken = this.shouldInvalidateAccessToken.bind(this);
+    this.invalidateAccessToken = this.invalidateAccessToken.bind(this);
+    this.handleUnauthorizedRequest = this.handleUnauthorizedRequest.bind(this);
+  }
+
+  /**
+   * Configures fetch interceptor with given config object
+   * @param initConfig
+   */
+  configure(initConfig) {
+    this.config = { ...this.config, ...initConfig };
+
+    if (!this.isConfigValid(this.config)) {
+      throw new Error(ERROR_INVALID_CONFIG);
+    }
+
+    this.accessTokenProvider = new AccessTokenProvider(this.fetch, this.config);
+  }
+
+  /**
+   * Authorizes fetch interceptor with given refresh token
+   * @param refreshToken
+   * @param accessToken
+   */
+  authorize(refreshToken, accessToken) {
+    this.accessTokenProvider.authorize(refreshToken, accessToken);
+  }
+
+  /**
+   * Returns current authorization for fetch fetchInterceptor
+   * @returns {{accessToken: string, refreshToken: string}}
+   */
+  getAuthorization() {
+    return this.accessTokenProvider.getAuthorization();
+  }
+
+  clear() {
+    this.accessTokenProvider.clear();
+  }
+
+  intercept(...args) {
+    return new Promise((resolve, reject) => this.resolveIntercept(resolve, reject, ...args));
+  }
+
+  resolveIntercept(resolve, reject, ...args) {
+    const request = new Request(...args);
+    const { accessToken } = this.accessTokenProvider.getAuthorization();
+
+    // if access token is not resolved yet
+    if (!accessToken) {
+      return this.accessTokenProvider
+        .refresh()
+        .then(() => this.fetchWithRetry(request, resolve, reject))
+        .catch(reject);
+    }
+
+    // attempt normal fetch operation
+    return this.fetchWithRetry(request, resolve, reject)
+      .catch(reject);
+  }
+
+  fetchWithRetry(request, outerResolve, outerReject) {
+    return Promise.resolve(this.createRequestUnit(request))
+      .then(this.shouldIntercept)
+      // authorize request
+      .then(this.authorizeRequest)
+      .then(this.shouldFetch)
+      // initial fetch
+      .then(this.fetchRequest)
+      .then(this.shouldInvalidateAccessToken)
+      .then(this.invalidateAccessToken)
+      .then(this.handleUnauthorizedRequest)
```

Comment:
handleUnauthorizedRequest hides that you are renewing token. Maybe this could be more explicit, by rejecting and then in rejectHandler try to resolve problem?